atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.thrift.TApplicationException;
+import org.junit.Test;
+
+public final class CassandraTopologyValidatorTest {
+
+    private static final CassandraServer NEW_HOST_ONE_CASSANDRA_SERVER =
+            CassandraServer.of(InetSocketAddress.createUnresolved("1.1.1.1", 80));
+    private static final String NEW_HOST_ONE = "new_host_one";
+    private static final String NEW_HOST_TWO = "new_host_two";
+    private static final String NEW_HOST_THREE = "new_host_three";
+    private static final Set<String> NEW_HOSTS = ImmutableSet.of(NEW_HOST_ONE, NEW_HOST_TWO, NEW_HOST_THREE);
+
+    private static final String OLD_HOST_ONE = "old_host_one";
+    private static final String OLD_HOST_TWO = "old_host_two";
+    private static final String OLD_HOST_THREE = "old_host_three";
+    private static final Set<String> OLD_HOSTS = ImmutableSet.of(OLD_HOST_ONE, OLD_HOST_TWO, OLD_HOST_THREE);
+    private static final Set<String> UUIDS = ImmutableSet.of("uuid1", "uuid2", "uuid3");
+
+    private static final Set<String> ALL_HOSTS = Sets.union(NEW_HOSTS, OLD_HOSTS);
+
+    private static final HostIdResult DEFAULT_RESULT = HostIdResult.success(UUIDS);
+
+    private final CassandraTopologyValidationMetrics metrics =
+            CassandraTopologyValidationMetrics.of(new DefaultTaggedMetricRegistry());
+    private CassandraTopologyValidator validator = spy(new CassandraTopologyValidator(metrics));
+
+    @Test
+    public void retriesUntilNoNewHostsReturned() {
+        Iterator<String> uuidIterator = UUIDS.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(NEW_HOSTS);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(
+                        Set.of(container),
+                        HostIdResult.success(Set.of(uuidIterator.next())),
+                        HostIdResult.success(UUIDS)));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(),
+                        allHosts,
+                        HumanReadableDuration.milliseconds(1),
+                        HumanReadableDuration.seconds(20)))
+                .isEmpty();
+    }
+
+    @Test
+    public void retriesThrowsAndMarksFailureWhenNeverResolves() {
+        Iterator<String> uuidIterator = UUIDS.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(NEW_HOSTS);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(Set.of(container), HostIdResult.success(Set.of(uuidIterator.next()))));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(),
+                        allHosts,
+                        HumanReadableDuration.milliseconds(1),
+                        HumanReadableDuration.milliseconds(1)))
+                .isNotEmpty();
```

Comment:
as @mdaudali notes: it's probably worth asserting the contents

Reply:
apologies missed your comment @mdaudali ! The reason why I explicitly didn't do this is b/c this is out-of-scope of the test; all we care about here are metrics being marked, and it'll be confusing later on if this test fails if the return type has changed (as the metrics may in fact be getting marked). I'll throw up a separate test to ensure we're returning exactly what `getNewHostsWithInconsistentTopologies` is returns, as I think that's what really is missing here.